File: backend/src/modules/catalog/domain/catalog.py
from dataclasses import dataclass, replace
from enum import StrEnum
from uuid import UUID

from modules.catalog.domain.exceptions import SelfRelationForbiddenError
from modules.catalog.domain.publication import ensure_publishable
from modules.catalog.domain.sources import Source


class PublicationStatus(StrEnum):
    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


class ResearchStatus(StrEnum):
    VERIFIED = "verified"
    NEEDS_REVIEW = "needs_review"


class EntityType(StrEnum):
    SETTLEMENT = "settlement"
    PERSON = "person"
    EVENT = "event"
    LANDMARK = "landmark"
    NATURAL_OBJECT = "natural_object"
    CULTURAL_OBJECT = "cultural_object"
    ORGANIZATION = "organization"
    UNIVERSITY_OBJECT = "university_object"
    ARTIFACT = "artifact"


class RelationType(StrEnum):
    BORN_IN = "born_in"
    LIVED_IN = "lived_in"
    WORKED_IN = "worked_in"
    STUDIED_IN = "studied_in"
    TAUGHT_AT = "taught_at"
    PARTICIPATED_IN = "participated_in"
    LOCATED_IN = "located_in"
    PART_OF = "part_of"
    CREATED_BY = "created_by"
    DESCRIBED_IN = "described_in"
    CONNECTED_WITH = "connected_with"
    CONNECTED_WITH_CHGU = "connected_with_chgu"


@dataclass(frozen=True, slots=True)
class CatalogEntity:
    id: UUID
    sources: tuple[Source, ...] = ()
    status: PublicationStatus = PublicationStatus.DRAFT

    def publish(self) -> "CatalogEntity":
        ensure_publishable(self.sources)
        return replace(self, status=PublicationStatus.PUBLISHED)


@dataclass(frozen=True, slots=True)
class CatalogRelation:
    id: UUID
    source_entity_id: UUID
    target_entity_id: UUID
    sources: tuple[Source, ...] = ()
    status: PublicationStatus = PublicationStatus.DRAFT

    def __post_init__(self) -> None:
        if self.source_entity_id == self.target_entity_id:
            raise SelfRelationForbiddenError

    def publish(self) -> "CatalogRelation":
        ensure_publishable(self.sources)
        return replace(self, status=PublicationStatus.PUBLISHED)

File: frontend/src/modules/admin-catalog/domain/catalog.ts
export type AdminEntityType =
  | "settlement"
  | "person"
  | "event"
  | "landmark"
  | "natural_object"
  | "cultural_object"
  | "organization"
  | "university_object"
  | "artifact";
export type AdminEntityStatus = "draft" | "published" | "archived";
export type AdminRelationType =
  | "born_in"
  | "lived_in"
  | "worked_in"
  | "studied_in"
  | "taught_at"
  | "participated_in"
  | "located_in"
  | "part_of"
  | "created_by"
  | "described_in"
  | "connected_with"
  | "connected_with_chgu";
export type AdminSourceType =
  | "archive_document"
  | "book"
  | "scientific_article"
  | "museum_material"
  | "official_publication"
  | "photo"
  | "audio"
  | "video"
  | "oral_testimony"
  | "web_resource";

export interface LocalizedText {
  ru: string;
  ce: string | null;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface AdminCatalogPermissions {
  read: boolean;
  write: boolean;
  export: boolean;
  auditRead: boolean;
}

export interface AdminEntityView {
  id: string;
  type: AdminEntityType;
  slug: string;
  title: LocalizedText;
  shortDescription: LocalizedText;
  fullDescription: LocalizedText;
  coordinates: Coordinates | null;
  periodFrom: number | null;
  periodTo: number | null;
  districtId: string | null;
  status: AdminEntityStatus;
  version: number;
  relationsCount: number;
  sourcesCount: number;
  mediaCount: number;
}

export interface EntityInput {
  type: AdminEntityType;
  slug: string;
  title: LocalizedText;
  shortDescription: LocalizedText;
  fullDescription: LocalizedText;
  coordinates: Coordinates | null;
  periodFrom: number | null;
  periodTo: number | null;
  districtId: string | null;
  status: Exclude<AdminEntityStatus, "archived">;
}

export interface EntityListFilters {
  query?: string;
  type?: AdminEntityType;
  status?: AdminEntityStatus;
  limit?: number;
  offset?: number;
}

export interface AdminRelationView {
  id: string;
  sourceEntityId: string;
  targetEntityId: string;
  type: AdminRelationType;
  title: LocalizedText;
  description: LocalizedText;
  periodFrom: number | null;
  periodTo: number | null;
  status: AdminEntityStatus;
  version: number;
}

export interface RelationInput {
  sourceEntityId: string;
  targetEntityId: string;
  type: AdminRelationType;
  title: LocalizedText;
  description: LocalizedText;
  periodFrom: number | null;
  periodTo: number | null;
  status: Exclude<AdminEntityStatus, "archived">;
}

export interface RelationListFilters {
  entityId?: string;
  type?: AdminRelationType;
  limit?: number;
  offset?: number;
}

export interface AdminSourceView {
  id: string;
  title: string;
  type: AdminSourceType;
  author: string | null;
  publisher: string | null;
  publicationYear: number | null;
  url: string | null;
  archiveReference: string | null;
  description: string;
  isVerified: boolean;
  status: AdminEntityStatus;
  version: number;
}

export interface SourceInput {
  title: string;
  type: AdminSourceType;
  author: string | null;
  publisher: string | null;
  publicationYear: number | null;
  url: string | null;
  archiveReference: string | null;
  description: string;
  isVerified: boolean;
  status: Exclude<AdminEntityStatus, "archived">;
}

export interface SourceListFilters {
  query?: string;
  type?: AdminSourceType;
  limit?: number;
  offset?: number;
}

export interface AuditView {
  id: string;
  actorId: string;
  action: string;
  resourceType: string;
  resourceId: string;
  resourceVersion: number;
  createdAt: string;
}

export interface BoundedPage<T> {
  items: T[];
  meta: { limit: number; offset: number; total: number };
}

export class AdminCatalogError extends Error {
  constructor(
    readonly code:
      | "bad_request"
      | "unauthorized"
      | "forbidden"
      | "conflict"
      | "not_found"
      | "export_too_large"
      | "source_required"
      | "validation_error"
      | "internal_error"
      | "service_unavailable",
    message: string,
  ) {
    super(message);
    this.name = "AdminCatalogError";
  }
}
